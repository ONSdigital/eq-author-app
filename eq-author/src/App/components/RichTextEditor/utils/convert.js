import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
