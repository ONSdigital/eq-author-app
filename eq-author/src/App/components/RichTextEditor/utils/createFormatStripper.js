import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
