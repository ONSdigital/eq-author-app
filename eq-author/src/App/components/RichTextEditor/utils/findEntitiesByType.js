export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
