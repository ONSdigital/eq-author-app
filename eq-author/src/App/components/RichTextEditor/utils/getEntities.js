function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
