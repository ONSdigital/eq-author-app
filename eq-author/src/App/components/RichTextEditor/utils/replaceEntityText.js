import { SelectionState, Modifier } from "draft-js";

const getEntityRange = (block, entityKey) => {
  let range;

  // get the range for the entity key
  block.findEntityRanges(
    char => char.getEntity() === entityKey,
    (start, end) => {
      range = { start, end };
    }
  );

  return SelectionState.createEmpty(block.key).merge({
    anchorKey: block.key,
    anchorOffset: range.start,
    focusKey: block.key,
    focusOffset: range.end
  });
};

const replaceText = (contentState, entityKey, blockKey, text, inlineStyles) => {
  const block = contentState.getBlockForKey(blockKey);
  const entitySelection = getEntityRange(block, entityKey);

  // first replace text
  const updatedContentState = Modifier.replaceText(
    contentState,
    entitySelection,
    text,
    inlineStyles,
    entityKey
  );

  // then create new selection, which covers the updated entity
  const updatedEntitySelection = entitySelection.merge({
    focusOffset: entitySelection.getStartOffset() + text.length
  });

  // apply the updated entity to the content
  return Modifier.applyEntity(
    updatedContentState,
    updatedEntitySelection,
    entityKey
  );
};

export default replaceText;
