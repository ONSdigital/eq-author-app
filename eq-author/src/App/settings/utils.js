export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
