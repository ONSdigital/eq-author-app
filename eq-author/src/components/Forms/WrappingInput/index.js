import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { radius } from "constants/theme";
import AutoResizeTextArea from "react-textarea-autosize";
import withChangeHandler from "components/Forms/withChangeHandler";
import { invoke } from "lodash";
import { sharedStyles } from "components/Forms/css";
import ValidationError from "components/ValidationError";

const ENTER_KEY = 13;

const StyleContext = styled.div`
  font-weight: ${(props) => (props.bold ? "bold" : "regular")};
`;

const TextArea = styled(AutoResizeTextArea)`
  ${sharedStyles};
  font-weight: inherit;
  border-radius: ${radius};
  resize: none;
  overflow: hidden; /* prevent scrollbars on Windows */
`;

class WrappingInput extends React.Component {
  static propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onKeyDown: PropTypes.func,
    id: PropTypes.string.isRequired,
    bold: PropTypes.bool,
    placeholder: PropTypes.string,
    errorValidationMsg: PropTypes.string,
  };

  static defaultProps = {
    bold: false,
  };

  handleChange = (e) => {
    e.target.value = e.target.value.replace(/\n/g, " ");
    this.props.onChange(e);
  };

  handleKeyDown = (e) => {
    if (e.keyCode === ENTER_KEY) {
      e.preventDefault();
    }

    invoke(this.props, "onKeyDown", e);
  };

  render() {
    const { bold, placeholder, errorValidationMsg, ...otherProps } = this.props;
    return (
      <StyleContext
        bold={bold}
        onChange={this.handleChange}
        onKeyDown={this.handleKeyDown}
      >
        <TextArea
          {...otherProps}
          placeholder={placeholder}
          invalid={errorValidationMsg}
          aria-invalid={Boolean(errorValidationMsg).toString()}
        />
        {errorValidationMsg && (
          <ValidationError>{errorValidationMsg}</ValidationError>
        )}
      </StyleContext>
    );
  }
}

export default withChangeHandler(WrappingInput);
