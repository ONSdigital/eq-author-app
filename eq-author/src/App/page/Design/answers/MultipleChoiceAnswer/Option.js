import React, { Component } from "react";
import styled from "styled-components";
import { colors, radius } from "constants/theme";
import { Field, Label } from "components/Forms";
import WrappingInput from "components/Forms/WrappingInput";
import withEntityEditor from "components/withEntityEditor";
import withValidationError from "enhancers/withValidationError";
import { flowRight, lowerCase } from "lodash";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import DeleteButton from "components/buttons/DeleteButton";
import Tooltip from "components/Forms/Tooltip";
import MoveButton, { IconUp, IconDown } from "components/buttons/MoveButton";
import { CHECKBOX, RADIO } from "constants/answer-types";
import DummyMultipleChoice from "../dummy/MultipleChoice";

import optionFragment from "graphql/fragments/option.graphql";
import getIdForObject from "utils/getIdForObject";

const ENTER_KEY = 13;

export const ButtonsContainer = styled.div`
  position: absolute;
  right: 0;
  top: 0;
  display: flex;
  z-index: 2;
  justify-content: flex-end;
`;

export const Flex = styled.div`
  display: flex;
  align-items: flex-start;
`;

export const OptionField = styled(Field)`
  margin-bottom: 1.5em;
`;

export const StyledOption = styled.div`
  border: 1px solid ${colors.bordersLight};
  padding: 1em 1em 0;
  border-radius: ${radius};
  position: relative;

  margin-bottom: 1em;
`;

StyledOption.defaultProps = {
  duration: 200,
};

StyledOption.propTypes = {
  duration: PropTypes.number,
};

export class StatelessOption extends Component {
  static propTypes = {
    option: CustomPropTypes.option.isRequired,
    onChange: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    onEnterKey: PropTypes.func,
    hasDeleteButton: PropTypes.bool.isRequired,
    type: PropTypes.oneOf([RADIO, CHECKBOX]).isRequired,
    children: PropTypes.node,
    labelPlaceholder: PropTypes.string,
    descriptionPlaceholder: PropTypes.string,
    autoFocus: PropTypes.bool,
    label: PropTypes.string,
    getValidationError: PropTypes.func,
    canMoveUp: PropTypes.bool,
    onMoveUp: PropTypes.func,
    canMoveDown: PropTypes.bool,
    onMoveDown: PropTypes.func,
    hideMoveButtons: PropTypes.bool,
  };

  static defaultProps = {
    labelPlaceholder: "Label",
    autoFocus: true,
    getValidationError: () => {},
  };

  handleDeleteClick = () => {
    this.props.onDelete(this.props.option.id);
  };

  handleKeyDown = e => {
    if (e.keyCode === ENTER_KEY) {
      this.props.onEnterKey(e);
    }
  };

  renderToolbar() {
    const {
      hideMoveButtons,
      hasDeleteButton,
      canMoveUp,
      onMoveUp,
      canMoveDown,
      onMoveDown,
    } = this.props;

    return (
      <ButtonsContainer>
        {!hideMoveButtons && (
          <>
            <Tooltip
              content="Move answer up"
              place="top"
              offset={{ top: 0, bottom: 10 }}
            >
              <MoveButton
                disabled={!canMoveUp}
                onClick={onMoveUp}
                data-test="btn-move-answer-up"
              >
                <IconUp />
              </MoveButton>
            </Tooltip>
            <Tooltip
              content="Move answer down"
              place="top"
              offset={{ top: 0, bottom: 10 }}
            >
              <MoveButton
                disabled={!canMoveDown}
                onClick={onMoveDown}
                data-test="btn-move-answer-down"
              >
                <IconDown />
              </MoveButton>
            </Tooltip>
          </>
        )}
        {(hasDeleteButton || !hideMoveButtons) && (
          <Tooltip content="Delete option" place="top" offset={{ bottom: 10 }}>
            <DeleteButton
              size="medium"
              aria-label="Delete option"
              onClick={this.handleDeleteClick}
              data-test="btn-delete-option"
              disabled={!hasDeleteButton}
            />
          </Tooltip>
        )}
      </ButtonsContainer>
    );
  }

  render() {
    const {
      option,
      onChange,
      onUpdate,
      type,
      children,
      labelPlaceholder,
      descriptionPlaceholder,
      autoFocus,
      label,
      getValidationError,
    } = this.props;

    return (
      <StyledOption id={getIdForObject(option)} key={option.id}>
        <div>
          <Flex>
            <DummyMultipleChoice type={type} />
            <OptionField>
              <Label htmlFor={`option-label-${option.id}`}>
                {label || "Label"}
              </Label>
              <WrappingInput
                id={`option-label-${option.id}`}
                name="label"
                value={option.label}
                placeholder={labelPlaceholder}
                onChange={onChange}
                onBlur={onUpdate}
                onKeyDown={this.handleKeyDown}
                data-test="option-label"
                data-autofocus={autoFocus || null}
                bold
                errorValidationMsg={getValidationError({
                  field: "label",
                  label: "Option label",
                  requiredMsg: `Enter a ${lowerCase(type)} label`,
                })}
              />
            </OptionField>
          </Flex>
          <OptionField>
            <Label htmlFor={`option-description-${option.id}`}>
              Description (optional)
            </Label>
            <WrappingInput
              id={`option-description-${option.id}`}
              name="description"
              placeholder={descriptionPlaceholder}
              onChange={onChange}
              value={option.description}
              onBlur={onUpdate}
              onKeyDown={this.handleKeyDown}
              data-test="option-description"
            />
          </OptionField>
          {children}
          {this.renderToolbar()}
        </div>
      </StyledOption>
    );
  }
}

StatelessOption.fragments = {
  Option: optionFragment,
};

export default flowRight(
  withValidationError("option"),
  withEntityEditor("option")
)(StatelessOption);
