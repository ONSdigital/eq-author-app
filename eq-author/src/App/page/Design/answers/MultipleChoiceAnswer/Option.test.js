import React from "react";
import WrappingInput from "components/Forms/WrappingInput";
import { shallow, mount } from "enzyme";
import { StatelessOption } from "./Option";
import { CHECKBOX, RADIO } from "constants/answer-types";
import { merge } from "lodash";
import { useMutation } from "@apollo/react-hooks";
import { props } from "lodash/fp";
import { render as rtlRender, fireEvent, screen } from "tests/utils/rtl";

jest.mock("@apollo/react-hooks", () => ({
  useMutation: jest.fn(),
}));

useMutation.mockImplementation(jest.fn(() => [jest.fn()]));

describe("Option", () => {
  let mockMutations;
  let mockEvent;
  let wrapper;

  const option = {
    id: "1",
    label: "",
    description: "",
    additionalAnswer: {
      id: "additional1",
      label: "",
      type: "TextField",
    },
    __typename: "Option",
  };

  const render = (method = shallow, otherProps) => {
    wrapper = method(
      <StatelessOption
        {...mockMutations}
        option={option}
        hasDeleteButton
        type={RADIO}
        {...otherProps}
      />
    );

    return wrapper;
  };

  beforeEach(() => {

    mockEvent = {
      stopPropagation: jest.fn(),
      preventDefault: jest.fn(),
    };

    mockMutations = {
      onBlur: jest.fn(),
      onChange: jest.fn(),
      onUpdate: jest.fn(),
      onFocus: jest.fn(),
      onDelete: jest.fn(),
      onEnterKey: jest.fn(),
    };

    render();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a checkbox", () => {
    render(mount, { type: CHECKBOX });
    expect(wrapper).toMatchSnapshot();
  });

  it("shouldn't render delete button if not applicable", () => {
    wrapper.setProps({ hasDeleteButton: false });
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onChange and onBlur correctly", () => {
    const { getByTestId } = rtlRender(( otherProps) => <StatelessOption {...mockMutations}
      option={option}
      hasDeleteButton
      type={RADIO}
      {...otherProps}
     {...props} />);

    fireEvent.change(getByTestId("option-label"), {
      target: { value: "2" },
    });
    fireEvent.blur(getByTestId("option-label"));
    expect(mockMutations.onUpdate).toHaveBeenCalledTimes(1);
  });

  it("should call onChange on input", async() => {
    render(rtlRender, { type: CHECKBOX });
    fireEvent.change(screen.getByTestId("other-answer"), {
      target: { value: "use this text" },
      });

      expect(
        screen.getByText(/use this text/)
      ).toBeInTheDocument();
  });

  it("should update label on blur", () => {
    render(rtlRender, { type: CHECKBOX });
    fireEvent.blur(screen.getByTestId("option-label"), {
      target: { value: "2" },
    });
    expect(mockMutations.onUpdate).toHaveBeenCalledTimes(1);
  });

  it("should update Other Answer on blur", () => {
    const handleSaveOtherLabel = jest.fn();
    useMutation.mockImplementation(() => [
      handleSaveOtherLabel
    ])
    render(rtlRender, { type: CHECKBOX });
    fireEvent.change(screen.getByTestId("other-answer"), {
      target: { value: "2" },
    });
    fireEvent.blur(screen.getByTestId("other-answer"),);
    expect(handleSaveOtherLabel).toHaveBeenCalledTimes(1);

  });

  it("should invoke onDelete callback when option deleted", () => {
    render(rtlRender, { type: CHECKBOX });
    fireEvent.click(screen.getByTestId("btn-delete-option"),);

    expect(mockMutations.onDelete).toHaveBeenCalledWith(option.id);
  });

  it("should call onEnterKey when Enter key pressed", () => {
    wrapper
      .find(WrappingInput)
      .first()
      .simulate("keyDown", merge(mockEvent, { keyCode: 13 }));

    expect(mockMutations.onEnterKey).toHaveBeenCalledWith(mockEvent);
  });

  it("should not call onEnterKey when other keys are pressed", () => {
    wrapper
      .find(WrappingInput)
      .first()
      .simulate("keyDown", merge(mockEvent, { keyCode: 27 }));

    expect(mockMutations.onEnterKey).not.toHaveBeenCalled();
  });

  it("can turn off auto-focus", () => {
    wrapper = render(mount, { autoFocus: false });
    const input = wrapper
      .find(`[data-test="option-label"]`)
      .first()
      .getDOMNode();

    expect(input.hasAttribute("data-autofocus")).toBe(false);
  });
});
