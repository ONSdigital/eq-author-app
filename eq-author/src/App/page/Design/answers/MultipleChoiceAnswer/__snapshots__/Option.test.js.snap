// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option shouldn't render delete button if not applicable 1`] = `
<Option__StyledOption
  duration={200}
  id="Option1"
  key="1"
>
  <div>
    <Option__ButtonsContainer>
      <Tooltip
        content="Move answer up"
        offset={
          Object {
            "bottom": 10,
            "top": 0,
          }
        }
        place="top"
      >
        <MoveButton
          aria-label="Move option up"
          color="secondary"
          data-test="btn-move-answer-up"
          disabled={true}
          tabIndex={-1}
        >
          <Component />
        </MoveButton>
      </Tooltip>
      <Tooltip
        content="Move answer down"
        offset={
          Object {
            "bottom": 10,
            "top": 0,
          }
        }
        place="top"
      >
        <MoveButton
          aria-label="Move option down"
          color="secondary"
          data-test="btn-move-answer-down"
          disabled={true}
          tabIndex={-1}
        >
          <Component />
        </MoveButton>
      </Tooltip>
      <Tooltip
        content="Delete option"
        offset={
          Object {
            "bottom": 10,
          }
        }
        place="top"
      >
        <DeleteButton
          aria-label="Delete option"
          color="secondary"
          data-test="btn-delete-option"
          disabled={true}
          onClick={[Function]}
          size="medium"
          tabIndex={-1}
          type="button"
        />
      </Tooltip>
    </Option__ButtonsContainer>
    <Option__Flex>
      <MultipleChoice
        type="Radio"
      />
      <Option__OptionField>
        <Label
          bold={true}
          htmlFor="option-label-1"
        >
          Label
        </Label>
        <withChangeHandler(WrappingInput)
          bold={true}
          data-autofocus={true}
          data-test="option-label"
          errorValidationMsg=""
          id="option-label-1"
          name="label"
          onBlur={[Function]}
          onChange={[MockFunction]}
          onKeyDown={[Function]}
          placeholder=""
          value=""
        />
      </Option__OptionField>
    </Option__Flex>
    <Option__OptionField>
      <Label
        bold={true}
        htmlFor="option-description-1"
      >
        Description (optional)
      </Label>
      <withChangeHandler(WrappingInput)
        data-test="option-description"
        id="option-description-1"
        name="description"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        onKeyDown={[Function]}
        value=""
      />
    </Option__OptionField>
    <Option__LastOptionField>
      <Label
        bold={true}
        htmlFor="option-otherLabel-additional1"
      >
        Other Label
      </Label>
      <withChangeHandler(WrappingInput)
        data-test="other-answer"
        id="option-otherLabel-additional1"
        name="otherLabel"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        value=""
      />
    </Option__LastOptionField>
  </div>
</Option__StyledOption>
`;
